src/nspawn/nspawn.c

```diff
@@ -2162,6 +2163,57 @@ static int setup_boot_id(void) {
         return mount_nofollow_verbose(LOG_ERR, NULL, to, NULL, MS_BIND|MS_REMOUNT|MS_RDONLY|MS_NOSUID|MS_NOEXEC|MS_NODEV, NULL);
 }
 
+static int get_fuse_version(uint32_t *major, uint32_t *minor) {
+        int r;
+        ssize_t n;
+        _cleanup_close_ int fd;
+        _cleanup_(umount_and_rmdir_and_freep) char *dir;
+        _cleanup_free_ char *opts = NULL;
+        union {
+                char unstructured[FUSE_MIN_READ_BUFFER];
+                struct {
+                        struct fuse_in_header header;
+                        /* Don't use <linux/fuse.h>:`struct fuse_init_in` because a newer fuse.h might give
+                         * us a bigger struct than what an older kernel actually gives us, and that would
+                         * break our short-read checks. */
+                        struct {
+                                uint32_t major;
+                                uint32_t minor;
+                        } body;
+                } structured;
+        } request;
+
+        /* Get a FUSE handle. */
+        fd = open("/dev/fuse", O_RDWR);
+        if (fd < 0)
+                return log_warning_errno(errno, "Disabling FUSE: Failed to open /dev/fuse: %m");
+        r = mkdtemp_malloc("/tmp/nspawn-fuse-XXXXXX", &dir);
+        if (r < 0)
+                return log_warning_errno(r, "Disabling FUSE: Failed to create temporary directory: %m");
+        if (asprintf(&opts, "fd=%i,rootmode=40000,user_id=0,group_id=0", fd) < 0)
+                return log_oom();
+        if (mount("nspawn-fuse", dir, "fuse.nspawn", 0, opts) < 0)
+                return log_warning_errno(r, "Disabling FUSE: Failed to mount test handle: %m");
+
+        /* Read a request from the FUSE handle. */
+        n = read(fd, &request.unstructured, sizeof request);
+        if (n < 0)
+                return log_warning_errno(errno, "Disabling FUSE: Failed to read /dev/fuse: %m");
+        if ((size_t) n < sizeof request.structured.header ||
+            (size_t) n < request.structured.header.len)
+                return log_warning_errno(-1, "Disabling FUSE: Failed to read /dev/fuse: Short read");
+
+        /* Assume that the request is a FUSE_INIT request, and return the version information from it. */
+        if (request.structured.header.opcode != FUSE_INIT)
+                return log_warning_errno(-1, "Disabling FUSE: Initial request from /dev/fuse should have opcode=%i (FUSE_INIT), but has opcode=%"PRIu32,
+                                   FUSE_INIT, request.structured.header.opcode);
+        if (request.structured.header.len < sizeof request.structured)
+                return log_warning_errno(-1, "Disabling FUSE: Initial FUSE_INIT request from /dev/fuse is too short");
+        *major = request.structured.body.major;
+        *minor = request.structured.body.minor;
+        return 0;
```

Comment:
hmm logging is all over the place here. some fatal calls use LOG_WARN, some user LOG_ERR.

please clean this up.

generally, I am pretty sure we should not log beyond LOG_DEBUG if fuse is not available at all. that should not be considered an error at all. maybe just return 0 with a major/minor of zero in that case.

Reply:
They all use LOG_WARN? Where are you seeing LOG_ERR?

But sure, I can demote them to LOG_DEBUG.

Reply:
log_oom() uses LOG_ERR.

key is:
1. expected errors should log LOG_DEBUG at most (an expected error is that kernel might not support fuse)
2. unexpected errors should log LOG_WARNING if we ignore it and proceed
3. unexpected errors should log LOG_ERR if we fail things because of it.

Reply:
Got it about `log_oom()`.

Except for `open` or the version number being too low, I don't expect any of the calls to fail. So should I use LOG_WARN for them?

Reply:
yes, I agree, open() failing with some ERRNO_IS_NEG_DEVICE_ABSENT() and the version check are kinda the only "expected" failures, and hence should result in debug logging at most.

for the others LOG_WARN level would be best.

hence, my suggestion: make sure the function returns zero major/minor and >= 0 ret code in case of the two expected failures (so that they are not considered failures, but instead just "no data" success responses), with debug logging. And return negative errno failure with LOG_WARN log message for the other errors you might hit. And if everything is order, return filled in major/minor and >= 0 ret code.